state/factory/factory.go

```diff
@@ -434,19 +440,35 @@ func (sf *factory) WorkingSetAtHeight(ctx context.Context, height uint64, preact
 }
 
 // PutBlock persists all changes in RunActions() into the DB
-func (sf *factory) PutBlock(ctx context.Context, blk *block.Block) error {
+func (sf *factory) PutBlock(ctx context.Context, blk *block.Block) (err error) {
 	timer := sf.timerFactory.NewTimer("Commit")
-	defer timer.End()
+	var (
+		ws      *workingSet
+		isExist bool
+	)
+	defer func() {
+		timer.End()
+		if err != nil {
+			// abandon current workingset, and all pending workingsets beyond current height
+			ws.abandon()
+			h, _ := ws.Height()
+			abandonWorkingSets(sf.workingsets, h)
+		}
+	}()
 	producer := blk.PublicKey().Address()
 	if producer == nil {
 		return errors.New("failed to get address")
 	}
 	ctx = protocol.WithRegistry(ctx, sf.registry)
 	key := generateWorkingSetCacheKey(blk.Header, blk.Header.ProducerAddress())
-	ws, isExist, err := sf.getFromWorkingSets(ctx, key)
+	ws, isExist, err = sf.getFromWorkingSets(ctx, key)
 	if err != nil {
-		return err
+		return
+	}
+	if err = ws.verifyParent(); err != nil {
+		return
 	}
+	ws.detachParent()
```

Comment:
this is used by `CommitBlock()`, at this time parent of current `ws` must have successfully committed his state changes into stateDB, so the current `ws` must only rely on itself when processing the block